patroni/postgresql/slots.py

```diff
@@ -33,6 +33,19 @@ def load_replication_slots(self):
             self._replication_slots = replication_slots
             self._schedule_load_slots = False
 
+    def ignore_replication_slot(self, cluster, slot):
+        ignore_slot = False
+        for matcher in cluster.config.ignore_slots_matchers:
+            if (
+                    (matcher.get("name") is None or matcher["name"] == slot["name"])
+                    and (matcher.get("database") is None or matcher["database"] == slot["database"])
+                    and (matcher.get("plugin") is None or matcher["plugin"] == slot["plugin"])
+                    and (matcher.get("type") is None or matcher["type"] == slot["type"])
+               ):
+                ignore_slot = True
+                break
+        ignore_slot
```

Comment:
The `slot` doesn't always have plugin and database keys, but only for logical slots, and that's the reason why tests sometimes fail.

```suggestion
    def ignore_replication_slot(self, cluster, name):
        slot = self._replication_slots[name]
        for matcher in cluster.config.ignore_slots_matchers:
            if all(not matcher.get(a) or matcher[a] == slot.get(a) for a in ('name', 'database', 'plugin', 'type')):
                return True
        return False
```

Reply:
It must be the case that some DCSs (etcd3 and kubernettes) handle errors differently. We consistently have the raft suite pass locally, but we noticed that actually the key errors that are logged in the Travis builds that failed are also logged when running the tests against raft even though they don't result in tests failing.

Looking into that further we realized that when we refactored out the slot check into a method we added a bug -- the slot name isn't present in the slot dict, so a key error was being thrown on basically every invocation (if `ignore_slots` had any config). Our tests passed though with raft because that error was being swallowed, and all of the tests were about slots being ignored. On the other hand etcd3 and kubernetes appear to stop doing anything when that key error occurs, which results in the failing test since Patroni never actually gets to the code that configured the state of the server.

We're pushing a fix that handles this issue plus updates the tests to guarantee it.